File: solutions/0001-two-sum/Solution.kt
class Solution {
    fun twoSum(nums: IntArray, target: Int): IntArray {
        val prevs = HashMap<Int, Int>()

        for ((index, x) in nums.withIndex()) {
            val indexOfCounterpart = prevs[target - x]
            if (indexOfCounterpart != null) {
                return intArrayOf(index, indexOfCounterpart)
            }

            prevs[x] = index
        }

        throw RuntimeException("no solution")
    }
}

File: solutions/0001-two-sum/solution.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> prevs;

        for (int i = 0; i < nums.size(); ++i) {
            auto counterpart = prevs.find(target - nums[i]);
            if (counterpart != prevs.end()) {
                return {counterpart->second, i};
            }
            prevs[nums[i]] = i;
        }

        throw logic_error("Problem statement promised a solution!");
    }
};
